### PR TITLE
Homework 1

### PR DESCRIPTION
**Pull request by mfedorchuck**
Homework 1

### CODE REVIEW

**Comment by mfedorchuck** [comment]
@v-hunter

File: data_transformer/main.py [hunk 1]
```diff
@@ -1,18 +1,36 @@
-from data_saver import get_exchange_rate
-from common import notify_tg
-from data_to_sheet import write_data
+import datetime
+from flask import Flask, request
+from flask import typing as flask_typing
 
+from data_transformer import transform
 
-target_note = get_exchange_rate(bank='mono')
```
Comment (mfedorchuck) [inline data_transformer/main.py:6]: Тут - тестував іншу відкриту апішку (https://api.monobank.ua/bank/currency)
P.S. у якості прикладу роботи з API з відкритими данними можна використовувати відкриті данні Приватбанку(у них є історія курсів валют за певний період)

File: data_saver/store.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+def clear_data_at_directory(path):
+    for filename in os.listdir(path):
+        file_path = os.path.join(path, filename)
+        try:
+            if os.path.isfile(file_path) or os.path.islink(file_path):
+                os.unlink(file_path)
+            elif os.path.isdir(file_path):
+                shutil.rmtree(file_path)
+        except Exception as e:
```
Comment (v-hunt) [inline data_saver/store.py:14]: When you raise the base exception, it's an antipattern called "Error hiding" ([link](https://en.wikipedia.org/wiki/Error_hiding))

If you're not going to use this exception in your code above, just don't raise it. If you're going to use it somehow, give this exception a name, then.

File: data_saver/store.py [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+def save_local(path: str, data: list) -> str:
+    print(path)
+    os.makedirs(name=path, exist_ok=True)
+    clear_data_at_directory(path)
+
+    with open(f'{path}/data.json', 'w') as f:
+        json.dump(data, f)
+
+    return "Data retrieved from API and stored successfully"
```
Comment (v-hunt) [inline data_saver/store.py:26]: A result should be returned by a function. Or something that indicates that a function ran successfully (e.g. boolean value).

I would just return nothing here (None)

File: data_saver/tests/test_api.py [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+    def test_api(self, get_batch_sales_data: mock.MagicMock):
+        """
+        Test with mocked service call
+        """
+        get_batch_sales_data.return_value = [{"client": "Bob Marley", "Profit": 12}]
+        respond = api.get_sales("http://test_url", "test_token", "2022-01-01")
+
+        self.assertEqual(type(respond), type([]))
+
+        print('✅ API test Ok')
```
Comment (v-hunt) [inline data_saver/tests/test_api.py:21]: No need to use print here. The test runner prints the message

File: data_saver/tests/test_api.py [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+
+    @mock.patch('data_saver.api.get_batch_sales_data')
+    def test_api(self, get_batch_sales_data: mock.MagicMock):
+        """
+        Test with mocked service call
+        """
+        get_batch_sales_data.return_value = [{"client": "Bob Marley", "Profit": 12}]
+        respond = api.get_sales("http://test_url", "test_token", "2022-01-01")
+
+        self.assertEqual(type(respond), type([]))
```
Comment (v-hunt) [inline data_saver/tests/test_api.py:19]: Some notes:
1. `type([])` is `list`
2. Many people test return list length as well. 
3. If you need to make `get_batch_sales_data` to return different results when you call it several times in your test, you can use the [side effect](https://docs.python.org/3/library/unittest.mock.html#unittest.mock.Mock.side_effect)
4. Also, you can use `self.assertListEquel` to test a list

File: data_transformer/main.py [hunk 6]
```diff
@@ -1,18 +1,36 @@
...
-    pass
+    input_data: dict = request.json
+
+    if "raw_dir" not in input_data or "stg_dir" not in input_data:
+        return {
+                   "message": "No path received. 'raw_dir' and 'stg_dir' need to be specified",
+               }, 400
+    else:
+        raw_dir_path = f"{input_data['raw_dir']}"
+        stg_dir_path = f"{input_data['stg_dir']}"
```
Comment (v-hunt) [inline data_transformer/main.py:25]: Why do you f-stringify those 2 parameters? They're strings already.
You can just use:
```
raw_dir_path = finput_data['raw_dir']
stg_dir_path = input_data['stg_dir']
```

File: data_transformer/transform.py [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+    return parse_schema(schema)
+
+
+def transform_and_save_data(raw_dir_path: str, stg_dir_path: str) -> str:
+    """
+    Transforming .json data into .avro format and save it.
+    Return status
+    """
+    try:
+        with open(f'{raw_dir_path}/data.json', 'r') as raw_file:
```
Comment (v-hunt) [inline data_transformer/transform.py:46]: please, pay attention to the correct name of the data file I proposed in the home task description. It should contain:
1. data source name
2. the date we run the job for
I will explain later why we need such a name convention

**Review by v-hunt — APPROVED** [review]
Very good.
Good that you wrote tests and used type hints.
Also, I left some comments, tips and recommendations.

**Pull request merged**